File: 714-BestTimeToBuyAndSellStockWithTransactionFee/714-BestTimeToBuyAndSellStockWithTransactionFee.py
# Last updated: 4/29/2026, 10:28:34 AM
"""
hold: dp(i) = dp(i+1, holding)
buy: -prices[i] + dp(i+1, true)
sell: prices[i] - 2 + dp(i+1, false)
"""
class Solution:
    def maxProfit(self, prices: List[int], fee: int) -> int:
        @cache
        def dp(i, holding):
            # print(f"{i}, {holding}")
            if i >= len(prices):
                return 0
            
            skip = dp(i+1, holding)
            if holding:
                return max(skip, prices[i] - fee + dp(i+1, False))
            else:
                return max(skip, -prices[i] + dp(i+1, True))
        
        return dp(0, False)

File: 714-BestTimeToBuyAndSellStockWithTransactionFee/714-BestTimeToBuyAndSellStockWithTransactionFee.js
// Last updated: 4/29/2026, 10:28:30 AM
/**
 * @param {number[]} prices
 * @param {number} fee
 * @return {number}
 */
var maxProfit = function(prices, fee) {
    const n = prices.length;
    const memo = new Array(n).fill(null).map(() => new Array(2).fill(null));
    
    const dp = (i, holding) => {
        if(i >= prices.length) {
            return 0;
        }
        if(memo[i][holding] !== null) {
            return memo[i][holding];
        }
        
        const skip = dp(i + 1, holding);
        if(holding === 0) {
            memo[i][holding] = Math.max(skip, -prices[i] + dp(i + 1, 1));
            return memo[i][holding];
        } else {
            memo[i][holding] = Math.max(skip, prices[i] - fee + dp(i+1, 0));
            return memo[i][holding];
        }
    }
    
    return dp(0, 0);
};